src/postamble.js

```diff
@@ -175,7 +175,11 @@ Module['callMain'] = Module.callMain = function callMain(args) {
 #endif
       return;
     } else {
-      if (e && typeof e === 'object' && e.stack) Module.printErr('exception thrown: ' + [e, e.stack]);
+      var hasStack = e && typeof e === 'object' && e.stack;
+      if (hasStack) Module.printErr('exception thrown: ' + [e, e.stack]);
+      // In a shell environment, callMain is ran from a promise context. Exceptions thrown from
+      // promises don't set the shell's exit code, so call quit here to force it to be non-zero.
+      quitIfShell(1, hasStack ? '' : ('exception thrown: ' + e));
```

Comment:
I'm not sure we want to be calling into the module's exit code if exceptions are thrown from or propagated through the wasm module code. But i guess that depends on a couple of things:
1. Things that could be called here would be wasm functions registered with `atexit` and... what else?
2. under what circumstances are exceptions thrown for wasm?. The motivating case here is `abort()` in which case I think we'd not want to call the atexit handlers.